File: java/ReverseNodesInKGroup.java
public class ReverseNodesInKGroup {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (k <= 1) {
            return head;
        }
        ListNode temp = new ListNode(0);
        temp.next = head;
        head = temp;
        ListNode start = head;
        ListNode end = head;
        while (true) {
            for (int i = 0; i < k; ++i) {
                end = end.next;
                if (end == null) {
                    break;
                }
            }
            if (end == null) {
                break;
            }
            ListNode token = end.next;
            end = start.next;
            ListNode reverseNode = null;
            for (ListNode ptr = start.next; ptr != token; ) {
                ListNode oldNode = reverseNode;
                reverseNode = ptr;
                ptr = ptr.next;
                reverseNode.next = oldNode;
            }
            start.next = reverseNode;
            end.next = token;
            start = end;
        }
        return head.next;
    }
}

File: cc/reverse_nodes_in_k_group.cc
#include "base_include.h"
#include "list_node.h"

class ReverseNodesInKGroup {
public:
    ListNode *reverseKGroup(ListNode *head, int k) {
        ListNode *token = new ListNode(0);
        token->next = head;
        head = token;
<<<<<<< HEAD
        ListNode *tail = head;
        for ( ; ; ) {
=======
        for ( ; ; ) {
            ListNode *tail = head;
>>>>>>> cee0291085b1fc5c266c9f29c21db92032414aa3
            for (int i = 0; i < k && tail != NULL; ++i) {
                tail = tail->next;
            }
            if (NULL == tail) {
                break;
            }
<<<<<<< HEAD
            ListNode *prev = head;
            head = head->next;
            ListNode *post = tail->next;
            ListNode *rev_node = NULL;
            for (ListNode *ptr = head; ptr != post; ) {
            }
        }
    }
};
=======
            ListNode *post = tail->next;
            ListNode *prev = NULL;
            for (ListNode *ptr = head->next; ptr != post; ) {
                ListNode *temp = ptr->next;
                ptr->next = prev;
                prev = ptr;
                ptr = temp;
            }
            ListNode *next_head = head->next;
            head->next->next = post;
            head->next = tail;
            head = next_head;
        }
        head = token->next;
        delete token;
        return head;
    }
};

int main(int argc, char *argv[]) {
    return 0;
}
>>>>>>> cee0291085b1fc5c266c9f29c21db92032414aa3
